### DISCUSSION TITLE
Some issues regarding the creation of custom Sessions.

### DISCUSSION THREAD

**Discussion by wyhdil** [eclipse-theia/theia-cloud#253] [Q&A] [answered]
Hello, I previously asked some questions in the community regarding the [customization of sessions](https://github.com/eclipsesource/theia-cloud/discussions/243). I have a general understanding of the process, but I encountered some issues during the code implementation. I would like to ask for your assistance. 

![b216a45ee5e5583f7a51d92e116bf4f6](https://github.com/eclipsesource/theia-cloud/assets/69418800/a93fe212-14f6-452a-b571-5432354323e9)
I deduced from the project code that the launch method is likely used to create a session CR in the cluster. However, I observed that the session object undergoes a change before and after the execution of the `operation().create(session)`  [operation](https://github.com/eclipsesource/theia-cloud/blob/main/java/common/org.eclipse.theia.cloud.common/src/main/java/org/eclipse/theia/cloud/common/k8s/client/DefaultSessionResourceClient.java#L52). Prior to the create method execution, the value of 'limitsCpu' is 1, but after the create method execution, the value changes to null.

The `operation().create() `method here should be using the Kubernetes Java library for interacting with the cluster, right? In theory, I believe the create method shouldn't modify the values of session.spec. For example, the values for 'limitsCpu' and 'requestsCpu' should remain unchanged. However, based on the printed log information, it appears that the create method is indeed causing modifications to the values in session.spec, such as setting 'limitsCpu' and 'requestsCpu' to null. I would like to understand the reason behind this.

Specifically, my goal is to extend the CPU and memory fields in the session CR, allowing users to define the resources required for the IDE pod, instead of obtaining CPU and memory resources from the App Definition CR.

I would like to inquire about how to resolve the issue with the mentioned `create` method, and whether my implementation approach is correct. Thanks!

**Comment by jfaltermeier** [comment, marked as answer]
Hi, 

the Java library we are using to interact with kubernetes is https://github.com/fabric8io/kubernetes-client
The `.create()` method should send to object you created to the kubernetes API and returns the object that is now present on kubernetes side. So it looks like on kubernetes your information got lost. 

It looks like you have changed the SessionSpec to also have the resource limitations in them. 
Besides that you also have to do two more things:
You also have to increase the version in the annotations in org.eclipse.theia.cloud.common.k8s.resource.Session. 
And tell kubernetes itself about this change by adding a new version here: https://github.com/eclipsesource/theia-cloud-helm/blob/main/charts/theia-cloud-crds/templates/session-spec-resource.yaml

**Reply by wyhdil** [reply]
Yes, after modifying the session CRD in the cluster, the customization of CPU and memory resources was successful！
Thank you for your assistance. Appreciate it!
